library/Freckle/App/Kafka.hs

```diff
@@ -0,0 +1,130 @@
+
+import Blammo.Logging
+import Control.Lens (Lens', view)
+import Data.Aeson (ToJSON, encode)
+import Data.ByteString (toStrict)
+import qualified Data.List.NonEmpty as NE
+import Data.Pool (Pool)
+import qualified Data.Pool as Pool
+import qualified Data.Text as T
+import qualified Freckle.App.Env as Env
+import Kafka.Producer
+import UnliftIO.Async (async)
+import UnliftIO.Exception (throwString)
+import Yesod.Core.Lens
+import Yesod.Core.Types (HandlerData)
+
+envKafkaBrokerAddresses
+  :: Env.Parser Env.Error (NonEmpty BrokerAddress)
+envKafkaBrokerAddresses = Env.var
+    (Env.eitherReader readKafkaBrokerAddresses)
+    "KAFKA_BROKER_ADDRESSES"
+    mempty
+
+readKafkaBrokerAddresses :: String -> Either String (NonEmpty BrokerAddress)
+readKafkaBrokerAddresses t = case NE.nonEmpty $ T.splitOn "," $ T.pack t of
+  Just xs@(x NE.:| _)
+    | x /= "" -> Right $ BrokerAddress <$> xs
+  _ -> Left "Broker Address cannot be empty"
+
+data KafkaProducerPool
+  = NullKafkaProducerPool
+  | KafkaProducerPool (Pool KafkaProducer)
+
+class HasKafkaProducerPool env where
+  kafkaProducerPoolL :: Lens' env KafkaProducerPool
+
+instance HasKafkaProducerPool site => HasKafkaProducerPool (HandlerData child site) where
+  kafkaProducerPoolL = envL . siteL . kafkaProducerPoolL
+
+createKafkaProducerPool :: NonEmpty BrokerAddress
+  -> Int
+  -- ^ The number of stripes (distinct sub-pools) to maintain.
+  -- The smallest acceptable value is 1.
+  -> NominalDiffTime
+  -- ^ Amount of time for which an unused resource is kept open.
+  -- The smallest acceptable value is 0.5 seconds.
+  --
+  -- The elapsed time before destroying a resource may be a little
+  -- longer than requested, as the reaper thread wakes at 1-second
+  -- intervals.
+  -> Int
+  -- ^ Maximum number of resources to keep open per stripe.  The
+  -- smallest acceptable value is 1.
+  --
+  -- Requests for resources will block if this limit is reached on a
+  -- single stripe, even if other stripes have idle resources
+  -- available.
+  -> IO (Pool KafkaProducer)
+createKafkaProducerPool addresses = Pool.createPool mkProducer closeProducer
+  where
+      mkProducer =
+        either throw pure =<< newProducer (brokersList $ toList addresses)
+      throw err = throwString $ "Failed to open kafka producer: " <> show err
+
+produceKeyedOn
+  :: ( ToJSON value
+     , ToJSON key
+     , MonadLogger m
+     , MonadReader env m
+     , HasKafkaProducerPool env
+     , MonadUnliftIO m
+     )
+  => TopicName
+  -> NonEmpty value
+  -> (value -> key)
+  -> m ()
+produceKeyedOn prTopic values keyF = do
+  logDebugNS "kafka" $ "Producing Kafka events" :# ["events" .= values]
+  view kafkaProducerPoolL >>= \case
+    NullKafkaProducerPool -> pure ()
+    KafkaProducerPool producerPool -> do
+      errors <-
+        liftIO $
+          Pool.withResource producerPool $ \producer ->
+            produceMessageBatch producer $
+              toList $
+                mkProducerRecord <$> values
```

Comment:
So this is an interesting interface. You're choosing to be opinionated and only support sending a batch of multiple messages (the function takes `ToJSON value => NonEmpty value` and not `ToJSON value => value`). This makes sense ergonomics-wise because the single-message case can easily be accomplished with a singleton list. I wonder if there are any other trade-offs to consider with this design. Does the underlying library also only expose `produceMessageBatch`, or does it have `produceMessage` too?

Reply:
There is also a `produceMessage` but I am not really sure about the tradeoffs involved. Both seem to just dump things on the queue that backs `librdkafka` and don't block. It looks like there is a `produceMessage'` that does have some blocking semantics if we ever need that for single messages.

The handler for Math adaptive can take multiple answers so I just exposed the `batch` variety.

Reply:
👍 That's probably a smart idea

Reply:
`produceBatchedKeyedOnAsync` is a mouthful though...

Reply:
Ugh, I know, not a fan.

Reply:
I think I'm just going to stick with the non-batched naming. We can adjust if we ever need the other semantics.